### PR TITLE
PSP-7500 - UI/UX cleanup agreements create /edit

### PR DESCRIPTION
**Pull request by eddherrera**
image

image

image

### CODE REVIEW

File: source/backend/api/Areas/Acquisition/Controllers/AgreementController.cs [hunk 1]
```diff
@@ -55,25 +62,119 @@ public AgreementController(IAcquisitionFileService acquisitionService, IMapper m
...
+                User.GetUsername(),
+                DateTime.Now);
+
             var agreements = _acquisitionService.GetAgreements(id);
             return new JsonResult(_mapper.Map<IEnumerable<AgreementModel>>(agreements));
         }
 
         /// <summary>
-        /// Update the acquisition file agreements.
+        /// Get the acquisition file agreements.
```
Comment (FuriousLlama) [inline source/backend/api/Areas/Acquisition/Controllers/AgreementController.cs:77]: This is kind of odd. The service call is updating the acq file by adding an agreement. Why the doc change?

File: source/backend/api/Services/AcquisitionFileService.cs [hunk 2]
```diff
@@ -411,6 +411,28 @@ public IEnumerable<PimsAgreement> GetAgreements(long id)
...
+
+            ValidateAcquisitionFileStatus(acquisitionFileId, agreement.AgreementStatusTypeCode);
+
+            var newAgreement = _agreementRepository.AddAgreement(agreement);
+            _agreementRepository.CommitTransaction();
+
+            return newAgreement;
+        }
+
+        public PimsAgreement GetAgreementById(long acquisitionFileId, long agreementId)
```
Comment (FuriousLlama) [inline source/backend/api/Services/AcquisitionFileService.cs:427]: the acqfile id is not being used

File: source/backend/api/Services/AcquisitionFileService.cs [hunk 3]
```diff
@@ -425,39 +447,35 @@ public IEnumerable<PimsAgreement> SearchAgreements(AcquisitionReportFilterModel
...
 
-            var toBeUpdated = currentAgreements.Where(ca => agreements.Any(na => ca.AgreementId == na.AgreementId && !ca.IsEqual(na)));
-            var toBeDeleted = currentAgreements.Where(ca => !agreements.Any(na => ca.AgreementId == na.AgreementId));
+            var updatedAgreement = _agreementRepository.UpdateAgreement(agreement);
+            _agreementRepository.CommitTransaction();
 
-            foreach (var agreement in toBeUpdated)
-            {
-                var agreementStatus = Enum.Parse<AgreementStatusTypes>(agreement.AgreementStatusTypeCode);
-                if (!_statusSolver.CanEditOrDeleteAgreement(currentAcquisitionStatus, agreementStatus) && !_user.HasPermission(Permissions.SystemAdmin))
```
Comment (FuriousLlama) [inline source/backend/api/Services/AcquisitionFileService.cs:442]: is this validation no longer necessary for the agreement?

Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:442]: nvm. saw the call to a validate function on 457

File: source/backend/api/Solvers/AcquisitionStatusSolver.cs [hunk 4]
```diff
@@ -70,25 +70,12 @@ public bool CanEditTakes(AcquisitionStatusTypes? acquisitionStatus)
...
-                    canEdit = false;
-                    break;
-                default:
-                    canEdit = false;
-                    break;
-            }
-
+                AcquisitionStatusTypes.ACTIVE or AcquisitionStatusTypes.DRAFT => true,
+                AcquisitionStatusTypes.ARCHIV or AcquisitionStatusTypes.CANCEL or AcquisitionStatusTypes.CLOSED or AcquisitionStatusTypes.COMPLT or AcquisitionStatusTypes.HOLD => false,
+                _ => false,
```
Comment (FuriousLlama) [inline source/backend/api/Solvers/AcquisitionStatusSolver.cs:77]: nit: not sure is more readable

File: source/backend/tests/unit/api/Controllers/Acquisition/AgreementControllerTest.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    [Trait("category", "api")]
+    [Trait("group", "acquisition")]
+    [ExcludeFromCodeCoverage]
+    public class AgreementControllerTest
+    {
+        private Mock<IAcquisitionFileService> _service;
+        private AgreementController _controller;
+        private IMapper _mapper;
+
+        public AgreementControllerTest()
```
Comment (FuriousLlama) [inline source/backend/tests/unit/api/Controllers/Acquisition/AgreementControllerTest.cs:26]: not sure this test file is necessary. All the tests below are just executing the 'happy path'. Also, the controller does not seem to have any logic worth testing.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/add/AddAcquisitionAgreementContainer.test.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+const mockPostApi = {
+  error: undefined,
+  response: undefined,
+  execute: jest.fn(),
+  loading: false,
+};
+const onSuccess = jest.fn();
+
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+let viewProps: IUpdateAcquisitionAgreementViewProps | undefined;
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/add/AddAcquisitionAgreementContainer.test.tsx:23]: nit: if not used, is it necessary?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.test.tsx [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  it('renders as expected', async () => {
+    const { asFragment } = await setup();
+    expect(asFragment()).toMatchSnapshot();
+  });
+
+  it('does not display cancellation note by default', async () => {
+    const { queryByText } = await setup();
+
+    expect(queryByText(/Cancellation reason/i)).toBeNull();
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.test.tsx:60]: nit: it would be better if the test data is setup explicitly in the test to make it obv what is being tested.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+import SidebarFooter from '@/features/mapSideBar/shared/SidebarFooter';
+import { getCancelModalProps, useModalContext } from '@/hooks/useModalContext';
+import { IApiError } from '@/interfaces/IApiError';
+import { ApiGen_Concepts_Agreement } from '@/models/api/generated/ApiGen_Concepts_Agreement';
+
+import AcquisitionAgreementForm from '../form/AcquisitionAgreementForm';
+import { AcquisitionAgreementFormYupSchema } from '../form/AcquisitionAgreementFormYupSchema';
+import { AcquisitionAgreementFormModel } from '../models/AcquisitionAgreementFormModel';
+
+export interface IUpdateAcquisitionAgreementViewProps {
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx:15]: On the name of the component, usually views only display data while forms contain editable sections.
In this case it would be UpdateAcquisitionAgreementForm

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+  isLoading: boolean;
+  initialValues: AcquisitionAgreementFormModel | null;
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx:17]: kind of nit. The container does not necessarily know what the initial values should be given the form is the one that is responsible of putting the data in whatever way it needs.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+              const agreementSaved = await onSave(values.toApi());
+              if (agreementSaved) {
+                onSuccess();
+              }
+            } catch (e) {
+              if (axios.isAxiosError(e)) {
+                const axiosError = e as AxiosError<IApiError>;
+                onError && onError(axiosError);
+              }
+            } finally {
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx:64]: the form should not 'know' that the save is going through an axios call. Also, this should be done by the container instead.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementContainer.tsx [hunk 11]
```diff
@@ -1,47 +1,88 @@
...
+    }
+
+    if (acquisitionFileResponse) {
+      const solverStatus = new StatusUpdateSolver(acquisitionFileResponse);
+      setStatusUpdateSolver(solverStatus);
+    }
+  }, [acquisitionFileId, getAcquisition, getAgreements]);
+
+  const handleAgreementDeleted = async (agreementId: number) => {
+    if (acquisitionFileId) {
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementContainer.tsx:65]: sugestion: use isValidId given it also tests for id of zero

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementView.tsx [hunk 12]
```diff
@@ -1,148 +1,224 @@
...
+        }
+      >
+        {agreements.map((agreement, index) => (
+          <StyledAgreementBorder key={`agreement-section-${index}`}>
+            <Section
+              header={
+                <Row>
+                  <Col md={5}>{`Agreement ${++index}`}</Col>
+                  <Col md={7}>
+                    {agreement.agreementType !== null && (
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementView.tsx:71]: exists(agreement.agreementType) is safer than just checking for null

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementView.tsx [hunk 13]
```diff
@@ -1,148 +1,224 @@
...
+              <SectionField labelWidth="5" label="Legal survey plan">
+                {agreement.legalSurveyPlanNum}
+              </SectionField>
+              <SectionField labelWidth="5" label="Agreement type">
+                {agreement.agreementType?.description}
+              </SectionField>
+              <SectionField labelWidth="5" label="Agreement date">
+                {prettyFormatDate(agreement.agreementDate)}
+              </SectionField>
+              {agreement.agreementType?.id === ApiGen_CodeTypes_AgreementTypes.H0074 && (
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementView.tsx:157]: +1 for using the strongly typed codes!

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+  const { setModalContent, setDisplayModal } = useModalContext();
+  const { getOptionsByType, getByType } = useLookupCodeHelpers();
+
+  const agreementStatusOptions = getOptionsByType(API.AGREEMENT_STATUS_TYPES);
+  const agreementTypeOptions = getByType(API.AGREEMENT_TYPES);
+
+  const agreementStatusTypeCodeValue = getIn(formikProps.values, 'agreementStatusTypeCode');
+  const agreementTypeCodeValue = getIn(formikProps.values, 'agreementTypeCode');
+  const agreementCancellationNoteValue = getIn(formikProps.values, 'cancellationNote');
+  const agreementStatusTypeCodeTouched = getIn(formikProps.touched, 'agreementStatusTypeCode');
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:34]: nit: better to be explicit about types here given getIn just returns any

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+
+  useEffect(() => {
+    if (
+      agreementStatusTypeCodeValue !== ApiGen_CodeTypes_AgreementStatusTypes.CANCELLED &&
+      !!agreementCancellationNoteValue
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:39]: unsure if this is safe. If Id isValidId, if string isValidString, if object exists()

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+              .call(e.target.selectedOptions)
+              .map((option: HTMLOptionElement & number) => option.value)[0];
+            if (!!selectedValue && selectedValue !== ApiGen_CodeTypes_AgreementTypes.H0074) {
+              formikProps.setFieldValue('commencementDate', '');
+            }
+          }}
+        />
+      </SectionField>
+      <SectionField labelWidth="5" label="Agreement date">
+        <FastDatePicker field={'agreementDate'} formikProps={formikProps} />
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:100]: nit, no need for brackets on the field

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/models/AcquisitionAgreementFormModel.ts [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+  public purchasePrice: string | null = '';
+  public legalSurveyPlanNum: string | null = '';
+  public offerDate: string | null = '';
+  public expiryDateTime: string | null = '';
+  public signedDate: string | null = '';
+  public inspectionDate: string | null = '';
+  public agreementStatusTypeCode: string | null = 'DRAFT';
+  public agreementStatusTypeDescription: string | null = '';
+  public cancellationNote: string | null = '';
+  public isDraft: boolean | null = null;
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/models/AcquisitionAgreementFormModel.ts:23]: all these are marked as nullable, but never actually set to null. Should they just be set to string instead?

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/update/UpdateAcquisitionAgreementContainer.tsx [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+  const onCreateError = (e: AxiosError<IApiError>) => {
+    if (e?.response?.status === 409) {
+      setModalContent({
+        variant: 'error',
+        title: 'Error',
+        message: 'You can only have one offer with status of "Accepted".',
+        okButtonText: 'Close',
+        handleOk: () => setDisplayModal(false),
+      });
+      setDisplayModal(true);
```
Comment (FuriousLlama) [inline source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/update/UpdateAcquisitionAgreementContainer.tsx:54]: IMO the modal should happen on the view instead of the container

File: source/backend/api/Services/AcquisitionFileService.cs (hunk 3 above)
Comment (eddherrera) [reply source/backend/api/Services/AcquisitionFileService.cs:442]: if so, please just resolve the comment.

File: source/backend/api/Solvers/AcquisitionStatusSolver.cs (hunk 4 above)
Comment (eddherrera) [reply source/backend/api/Solvers/AcquisitionStatusSolver.cs:77]: I am favor to start introducing more new features.

File: source/backend/tests/unit/api/Controllers/Acquisition/AgreementControllerTest.cs (hunk 5 above)
Comment (eddherrera) [reply source/backend/tests/unit/api/Controllers/Acquisition/AgreementControllerTest.cs:26]: @devinleighsmith has requested this in the past to keep the test coverage compliance. It you think it's something not worth doing please bring it up in the next 'Continuous development' meeting.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/add/AddAcquisitionAgreementContainer.test.tsx (hunk 6 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/add/AddAcquisitionAgreementContainer.test.tsx:23]: removed

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx (hunk 8 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx:15]: Updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx (hunk 9 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx:17]: this comment makes no sense.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx (hunk 15 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:39]: not sure what you mean here. I haven't changed the logic from the previous implementation.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx (hunk 16 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:100]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/models/AcquisitionAgreementFormModel.ts (hunk 17 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/models/AcquisitionAgreementFormModel.ts:23]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx (hunk 14 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:34]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementContainer.tsx (hunk 11 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementContainer.tsx:65]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementView.tsx (hunk 12 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/detail/AgreementView.tsx:71]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx (hunk 9 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx:17]: What part of it lost you? I can rephrase it if necessary.
I summary, there could be a better separation of concerns in this code.

File: source/backend/api/Services/AcquisitionFileService.cs (hunk 2 above)
Comment (FuriousLlama) [reply source/backend/api/Services/AcquisitionFileService.cs:427]: Ah, didn't see the call below. Thanks for the clarification

File: source/backend/api/Solvers/AcquisitionStatusSolver.cs (hunk 4 above)
Comment (FuriousLlama) [reply source/backend/api/Solvers/AcquisitionStatusSolver.cs:77]: Newer is not necessarily better. If you believe this new style is better, please elaborate. If there is no performance improvements, changes should be aimed at increasing consistency and readability in the code. If new language features help there we should use them, otherwise not.

File: source/backend/api/Areas/Acquisition/Controllers/AgreementController.cs (hunk 1 above)
Comment (eddherrera) [reply source/backend/api/Areas/Acquisition/Controllers/AgreementController.cs]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.test.tsx (hunk 7 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.test.tsx]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx (hunk 10 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx]: updated

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx (hunk 9 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/common/UpdateAcquisitionAgreementView.tsx]: The view is the same for both create and update containers ergo the view receives a set values (model) depending on the container that it's being invoked. The form (view) is just a representation.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/update/UpdateAcquisitionAgreementContainer.tsx (hunk 18 above)
Comment (eddherrera) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/update/UpdateAcquisitionAgreementContainer.tsx:50]: I think it's ok to be on the container since it would need the Axios context to handle issues.

File: source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx (hunk 15 above)
Comment (FuriousLlama) [reply source/frontend/src/features/mapSideBar/acquisition/tabs/agreement/form/AcquisitionAgreementForm.tsx:48]: its about the double bang (!!) used here. Javascript has a bunch of quirks and it would be better if we move away from using them when possible. If this is just moved logic no need to update it.

**Review by FuriousLlama — APPROVED** [review]
